src/Language/Spectacle/Syntax/Modal/Preterm.hs

```diff
@@ -0,0 +1,316 @@
+-- this way lets us avoid having to deduce which modality to use in the model checker as it is immediately apparent at
+-- the root of a formula's AST. An example of the reduction process for the formula such as
+--
+-- @
+-- (□ p₁ ∧ □ p₂) ∧ ¬ (◇ p₃ ∨ ◇ p₄) ≡ □ (p₁ ∧ p₂ ∧ p₃ ∧ p₄)
+-- @
+--
+-- would be:
+--
+-- @
+--             (∧)
+--              │                       (∧)                     (□)
+--        ╭─────┴─────╮                  │                       │
+--       (∨)         (¬)   ====>    ╭────┴────╮     ====>       (∧)
+--        │           │            (∧)       (∧)           ╭─────┴─────╮
+--     ╭──┴──╮        │          ╭──┴──╮   ╭──┴──╮        (∧)         (∧)
+--    (□)   (□)      (∨)        (□)   (□) (□)   (□)     ╭──┴──╮     ╭──┴──╮
+--     │     │     ╭──┴──╮       │     │   │     │     p₁     p₂   p₃     p₄
+--     p₁    p₂   (◇)   (◇)      p₁    p₂  p₃    p₄
+--                 │     │
+--                 p₂    p₄
+-- @
+--
+--
+--
+-- @since 0.1.0.0
+module Language.Spectacle.Syntax.Modal.Preterm
+  ( Preterm
+      ( PreConst,
+        PreConjunct,
+        PreDisjunct,
+        PreImplies,
+        PreNotImplies,
+        PreComplement,
+        PreAlways,
+        PreUpUntil
+      ),
+    materialize,
+    abstract,
+    normalizePreterm,
+    rewritePreterm,
+  )
+where
+
+import Data.Function ((&))
+import Data.Void (absurd)
+
+import Data.Functor.Loom (hoist, runLoom, (~>~))
+import Language.Spectacle.Lang
+  ( Lang (Op, Pure, Scoped),
+    Members,
+    decomposeOp,
+    decomposeS,
+    scope,
+    weaken,
+  )
+import Language.Spectacle.Syntax.Logic.Internal (Effect (Complement, Conjunct, Disjunct), Logic (Logic))
+import Language.Spectacle.Syntax.Modal.Internal (Effect (Always, UpUntil), Modal (Modal))
+
+-- ---------------------------------------------------------------------------------------------------------------------
+
+-- | 'Preterm' is the raw underlying syntax tree representing a temporal formula modulo the continuation or
+-- 'Data.Functor.Loom' information. A tree of 'Preterm's can be built by temporarily dispatching the 'Modal' and 'Logic'
+-- effects with 'materialize' and converted back into their effectful representation with 'abstract', which forms an
+-- isomorphism:
+--
+-- @
+--                                                      'materialize'
+--                                   ───────────────────────────────────────────────>
+-- Members '[Modal, Logic] effs =>                                                      Preterm Bool
+--   Lang ctx effs Bool              <───────────────────────────────────────────────
+--                                                       'abstract'
+-- @
+--
+-- 'Preterm's are rewritten in the normalization process rather than the effects directly since interposing effect
+-- translations directly would be extremely complicated for the set of tautologies used to normalize temporal formula.
```

Comment:
```suggestion
-- translations directly would be extremely complicated for the set of tautologies used to normalize temporal formulas.
```

Reply:
I see your formulas and raise you a _formulae_ :p